<a href="https://colab.research.google.com/github/KilovoIt/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/Stepan_DS_18_LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [ ]:
# YOUR WORK HERE
import pandas as pd
!wget 'https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip'
!unzip student.zip
student = pd.read_csv('student-mat.csv', delimiter=";")


--2020-07-14 19:21:00--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.07s   

2020-07-14 19:21:00 (289 KB/s) - ‘student.zip’ saved [20478/20478]

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [ ]:
student.reset_index()

index school sex  age address famsize  ... Walc  health  absences  G1  G2  G3
0        0     GP   F   18       U     GT3  ...    1       3         6   5   6   6
1        1     GP   F   17       U     GT3  ...    1       3         4   5   5   6
2        2     GP   F   15       U     LE3  ...    3       3        10   7   8  10
3        3     GP   F   15       U     GT3  ...    1       5         2  15  14  15
4        4     GP   F   16       U     GT3  ...    2       5         4   6  10  10
..     ...    ...  ..  ...     ...     ...  ...  ...     ...       ...  ..  ..  ..
390    390     MS   M   20       U     LE3  ...    5       4        11   9   9   9
391    391     MS   M   17       U     LE3  ...    4       2         3  14  16  16
392    392     MS   M   21       R     GT3  ...    3       3         3  10   8   7
393    393     MS   M   18       R     LE3  ...    4       5         0  11  12  10
394    394     MS   M   19       U     LE3  ...    3       5         5   8   9   9

[395 rows x 34 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [ ]:
student['studytime'].value_counts()

2    198
1    105
3     65
4     27
Name: studytime, dtype: int64

In [ ]:
# YOUR WORK HERE
import numpy as np
from scipy import stats
from scipy.stats import chi2_contingency 




null hypothesis: activities and studytime are independent
alternative: activities and studytime are dependent

In [ ]:
contingency = pd.crosstab(student['activities'], student['studytime'])




In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)
print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 3.4895734879706772
p value 0.3221169243533421
degrees of freedom 3
expected frequencies table 
 [[ 51.56962025  97.24556962  31.92405063  13.26075949]
 [ 53.43037975 100.75443038  33.07594937  13.73924051]]


based on P-value 0.32 and statistic of 3.49 we fail to reject null that activities and studytime are independent. 

In [ ]:
contingency = pd.crosstab(student['activities'], student['studytime'])
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)
print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 3.4895734879706772
p value 0.3221169243533421
degrees of freedom 3
expected frequencies table 
 [[ 51.56962025  97.24556962  31.92405063  13.26075949]
 [ 53.43037975 100.75443038  33.07594937  13.73924051]]


In [ ]:
contingency = pd.crosstab(student['higher'], student['failures'])
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)
print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 45.21014899446959
p value 8.348431490698966e-10
degrees of freedom 3
expected frequencies table 
 [[ 15.79746835   2.53164557   0.86075949   0.81012658]
 [296.20253165  47.46835443  16.13924051  15.18987342]]


In [ ]:
contingency = pd.crosstab(student['activities'], student['nursery'])
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)
print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 0.004951472553918086
p value 0.9439017945961875
degrees of freedom 1
expected frequencies table 
 [[ 39.78227848 154.21772152]
 [ 41.21772152 159.78227848]]


### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [ ]:
# YOUR WORK HERE
observed = pd.crosstab(student['activities'], student['nursery'])
observed

nursery     no  yes
activities         
no          40  154
yes         41  160

In [ ]:
row_totals = student['nursery'].value_counts().sort_index().values
row_totals

array([ 81, 314])

In [ ]:
column_totals = student['activities'].value_counts().sort_index().values
column_totals

array([194, 201])

In [ ]:
sample_size = student.shape[0]
sample_size

395

In [ ]:
expected = np.array([[0.0,0.0], [0.0,0.0]])

for row_position, col in enumerate(column_totals):
    for col_position, row in enumerate(row_totals):
        expected_value = (row*col) / sample_size
        expected[row_position][col_position] = expected_value

expected

array([[ 39.78227848, 154.21772152],
       [ 41.21772152, 159.78227848]])

In [ ]:
chi2 = ((observed-expected)**2/expected).sum().sum()
chi2

0.0029456527184359934

In [ ]:
dof=1,
pvalue = 0.96

based on Pvalue = 0.96 and chi2 = 0.002, we can reject the null.

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)